Support BPF_PROG_{QUERY,ATTACH,DETACH}

multun opened the pull request:
Fixes #3652

rocallahan commented on src/test/bpf_query.c:
```diff
@@ -0,0 +1,107 @@
+      atomic_puts("EXIT-SUCCESS");
+      return 0;
+    }
+    test_assert(!"failed to create cgroup");
```
So how does this test work when you run multiple copies simultaneously? Can we create a unique cgroup per test?

rocallahan replied:
Or can we use socket or perf eBPF programs instead so it's simpler to set up the right environment?

rocallahan replied:
Or are cgroups the only way to test `BPF_PROG_ATTACH`?

multun replied:
> So how does this test work when you run multiple copies simultaneously? Can we create a unique cgroup per test?

I'll create a cgroup per test

> Or can we use socket or perf eBPF programs instead so it's simpler to set up the right environment?

I believe socket / perf programs use another API. Here the full list of attach types:
https://www.kernel.org/doc/html/latest/bpf/libbpf/program_types.html

Most are cgroup stuff and kernel probes. I wish one of the types could be attached to an easier to create fd, but can't find any. I believe setting up a kprobe would be slightly easier, but would also have concurrency issues.

rocallahan commented on src/record_syscall.cc:
```diff
@@ -2194,6 +2194,46 @@ static Switchable prepare_bpf(RecordTask* t,
+    case BPF_PROG_DETACH:
+    case BPF_PROG_ATTACH: {
+      auto attr_size = t->regs().arg3();
+      syscall_state.reg_parameter(2, attr_size, IN_OUT);
```
Do these syscalls actually modify memory in user space?

rocallahan commented on src/record_syscall.cc:
```diff
@@ -2194,6 +2194,46 @@ static Switchable prepare_bpf(RecordTask* t,
+      // for each output array, only track changes if the field is
+      // within the bounds of the user provided structure
+      if ((prog_ids_p + 1).as_int() <= end_addr)
+        syscall_state.mem_ptr_parameter(prog_ids_p, buf_size, OUT);
```
{} around if bodies

rocallahan replied:
`OUT` is the default so let's just leave that out

rocallahan commented on src/record_syscall.cc:
```diff
@@ -2194,6 +2194,46 @@ static Switchable prepare_bpf(RecordTask* t,
+      // the syscall will fail and we can't track anything
+      auto prog_cnt_p = REMOTE_PTR_FIELD(attrp, query.prog_cnt);
+      if ((prog_cnt_p + 1).as_int() > end_addr)
+        break;
```
{} around if body

multun commented:
I've just noticed that just running ```rr record runc --help; rr replay`` does not work on my laptop, with or without this PR. I'll until I can test on a processor where this works (I'm pretty sure this works fine on my workstation)

multun commented:
I've made a couple of changes:
 - I've added a remote_buffer class to make the bound checks more readable
 - I've changed the test so it uses a network namespace and a BPF flow dissector

rocallahan commented on src/record_syscall.cc:
```diff
@@ -2196,42 +2197,46 @@ static Switchable prepare_bpf(RecordTask* t,
     case BPF_PROG_ATTACH: {
       auto attr_size = t->regs().arg3();
-      syscall_state.reg_parameter(2, attr_size, IN_OUT);
+      syscall_state.reg_parameter(2, attr_size, IN);
```
We don't need this at all. You can just ignore it.

rocallahan commented on src/record_syscall.cc:
```diff
@@ -2196,42 +2197,46 @@ static Switchable prepare_bpf(RecordTask* t,
-        .cast<typename Arch::bpf_attr>();
-      auto end_addr = attrp.as_int() + attr_size;
+      auto attr_begin = syscall_state.reg_parameter(2, attr_size, IN_OUT);
+      auto attr_buf = remote_buffer<void>(attr_begin, attr_size);
```
How about just using `MemoryRange` instead of `remote_buffer`?

multun replied:
I... looked for it and couldn't find it :clown_face: thanks!

khuey commented:
It seems like we should probably be doing the thing bpf does where it checks that the space (if any) between sizeof(bpf_attr) and the size argument to the syscall is 0. Otherwise if these structs get extended underneath us on newer kernels rr will silently diverge.

multun commented:
@khuey I've implemented your suggestion in [041eba7](https://github.com/rr-debugger/rr/pull/3653/commits/041eba751a1cdb34342e30a38d405e6422a57d3d)

rocallahan commented on src/kernel_abi.h:
```diff
@@ -1781,6 +1781,29 @@ struct BaseArch : public wordsize,
+    struct { /* anonymous struct used by BPF_PROG_QUERY command */
+      union {
+        __u32   target_fd;      /* target object to query or ... */
+        __u32   target_ifindex; /* target ifindex */
```
Get rid of the extra spaces between the type and the field name, here and below.

rocallahan reviewed (changes requested):
Getting close... sorry for the iterations here.

I'd prefer it if you just squash all these commits together for review and merging.

rocallahan commented:
Thanks!

multun commented:
@khuey can you try this out? #3663
